pkg/controllers/management/clusterupstreamrefresher/eks_upstream_spec.go

```diff
@@ -0,0 +1,136 @@
+		&eks.DescribeClusterInput{
+			Name: aws.String(cluster.Spec.EKSConfig.DisplayName),
+		})
+	if err != nil {
+		return nil, err
+	}
+
+	ngs, err := eksService.ListNodegroups(
+		&eks.ListNodegroupsInput{
+			ClusterName: aws.String(cluster.Spec.EKSConfig.DisplayName),
+		})
+
+	// gather upstream node groups states
+	var nodeGroupStates []*eks.DescribeNodegroupOutput
+	var errs []string
+	for _, ngName := range ngs.Nodegroups {
+		ng, err := eksService.DescribeNodegroup(
+			&eks.DescribeNodegroupInput{
+				ClusterName:   aws.String(cluster.Spec.EKSConfig.DisplayName),
+				NodegroupName: ngName,
+			})
+		if err != nil {
+			return nil, err
+		}
+
+		nodeGroupStates = append(nodeGroupStates, ng)
+		var nodeGroupMustBeDeleted string
+		if len(ng.Nodegroup.Health.Issues) != 0 {
+			var issueMessages []string
+			for _, issue := range ng.Nodegroup.Health.Issues {
+				issueMessages = append(issueMessages, aws.StringValue(issue.Message))
+				if !ekscontroller.NodeGroupIssueIsUpdatable(aws.StringValue(issue.Code)) {
+					nodeGroupMustBeDeleted = ": node group cannot be updated, must be deleted and recreated"
+				}
+			}
+			errs = append(errs, fmt.Sprintf("health error for node group [%s] in cluster [%s]: %s%s",
+				aws.StringValue(ng.Nodegroup.NodegroupName),
+				cluster.Name,
+				strings.Join(issueMessages, "; "),
+				nodeGroupMustBeDeleted,
+			))
+		}
+	}
+
+	upstreamSpec, _, err := ekscontroller.BuildUpstreamClusterState(cluster.Spec.DisplayName, cluster.Status.EKSStatus.ManagedLaunchTemplateID, clusterState, nodeGroupStates, ec2.New(sess), false)
+	if err != nil {
+		// If we get an error here, then syncing is broken
+		return nil, err
+	}
+
+	upstreamSpec.DisplayName = cluster.Spec.EKSConfig.DisplayName
+	upstreamSpec.Region = cluster.Spec.EKSConfig.Region
+	upstreamSpec.AmazonCredentialSecret = cluster.Spec.EKSConfig.AmazonCredentialSecret
+	upstreamSpec.Imported = cluster.Spec.EKSConfig.Imported
+	upstreamSpec.Subnets = cluster.Spec.EKSConfig.Subnets
+	upstreamSpec.SecurityGroups = cluster.Spec.EKSConfig.SecurityGroups
+	upstreamSpec.ServiceRole = cluster.Spec.EKSConfig.ServiceRole
+
+	if len(errs) != 0 {
+		// If there are errors here, we can still sync, but there are problems with the nodegroups that should be reported
+		err = fmt.Errorf("error for cluster [%s]: %s",
+			cluster.Name,
+			strings.Join(errs, "\n"))
+	}
+
+	return upstreamSpec, err
+}
+
+// MigrateEksRefreshCronSetting migrates the deprecated eks-refresh-cron setting to new
+// setting only if default setting was changed
+// This function will be run only once during startup by pkg/multiclustermanager/app.go
+func MigrateEksRefreshCronSetting(wContext *wrangler.Context) {
+	settingsClient := wContext.Mgmt.Setting().Cache()
+	eksRefreshSetting, err := settingsClient.Get(eksUpstreamRefresh)
+	if errors.IsNotFound(err) || eksRefreshSetting.Value != "" {
+		return
+	} else if err != nil {
+		logrus.Errorf("Error getting %s setting: %v", eksUpstreamRefresh, err)
```

Comment:
The errors in this section need to be explicit on where they are coming from and what they mean. This is important since we only do this once on startup, so the above error would just print once but as an operator it won't mean anything. 
```suggestion
		logrus.Errorf("Unable to complete Eks cron migration, will attempt at next rancher startup. Error getting %s setting: %v", eksUpstreamRefresh, err)
```

Reply:
This is an example of wording but should convey the intent